[@types/graphql-resolvers] Allow passing arbitrary TArgs to all utilities

aecorredor opened the pull request:
I'm facing an issue when using `graphql-resolvers` along with `graphqlgen` (for generating resolver types: https://github.com/prisma/graphqlgen). I get correct generated types from `graphqlgen`, but when I try to create a resolver using `combineResolvers`, I get a type error.

Here is an example:
```js
contest: combineResolvers(authHelpers.isAuthenticated, contestResolver)
```

The snippet above gives me this typescript error:
```
Types of parameters 'args' and 'args' are incompatible.
    Property 'id' is missing in type 'TArgs' but required in type 'ArgsContest'.
```

Where my `contestResolver` has the type `(parent: undefined, args: QueryResolvers.ArgsContest, ctx: Context, info: GraphQLResolveInfo) => Contest | Promise<Contest>`.

Where `ArgsContests` is  
```
export interface ArgsContest {
    id: string;
}
```

And my `isAuthenticated` resolver has type `(parent: any, args: any, { me }: Context) => ForbiddenError | undefined`.

Since right now I have no way of passing my own `TArgs` I have no other way but to ignore the error.

With this PR I introduce the ability to pass a generic `TArgs` of my liking that let's me make typescript happy when using `graphqlgen` generated types. If no generic for that argument is passed, then `TArgsDefault` is used, which is equivalent to what the previous typings had.

Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Add or edit tests to reflect the change. (Run with `npm test`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:f changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes:  https://github.com/prisma/graphqlgen

aecorredor commented:
@mike-engel hey man, if you have time, could you help me out in figuring out why the travis build is failing? originally I thought it was because of the way I was modifying the contributors, but apparently that's not it. The error message in the logs is not helpful at all either. I looked in another issue and they were getting the same error message when the contributors format was wrong. Thanks!

aecorredor commented:
I'm also trying to run the linter by following the instructions on the repo, but it keeps failing because it doesn't find `graphql-tools`. Have you faced this issue before? Thanks.

mike-engel commented on types/graphql-resolvers/index.d.ts:
```diff
@@ -1,37 +1,38 @@
 // Project: https://github.com/lucasconstantino/graphql-resolvers#readme
-// Definitions by: Mike Engel <https://github.com/mike-engel>
+// Definitions by: Mike Engel <https://github.com/mike-engel>, Alejandro Corredor <https://github.com/aecorredor>
+
```
Looks like you added an extra line here which might be breaking the header section. Additionally, looking at the [mongodb](https://github.com/smeijer/DefinitelyTyped/blob/a5f93e601f02d025a48ffb2dabd93b4bd9163296/types/mongodb/index.d.ts#L4) types, it looks like maybe each contributor should be on their own line.

aecorredor commented:
> LGTM. I think I've seen those errors before in the tests. Do the tests run ok for you locally?

You mean the linter or the tests? I couldn't run the linter because of that error related to not finding the `graphql-tools` module, even though I had installed all dependencies. The tests I'll try to run right now.

But yeah, it seems to be that extra line breaking it. Let me fix that, run tests, and see if the build passes. Thanks!

mike-engel commented:
@aecorredor Sorry, I meant the linter tests due to the missing `graphql-tools`. The error I actually saw in the travis logs looked different than the ones I've seen before

aecorredor commented:
Yeah I think the one from the travis logs is related to the error with the contributors format. The linter I haven't been able to figure out.

aecorredor commented:
@mike-engel so now the actual linter is failing: https://travis-ci.org/DefinitelyTyped/DefinitelyTyped/builds/566005261?utm_source=github_status&utm_medium=notification, mostly due to external dependencies failing tslint. Is this what you have run into before? I'm not sure how to proceed here.

mike-engel commented:
@aecorredor Yeah, that looks like an error with code that's not yours. Looking at other PRs, however, shows that most are passing the tests. It's possible the `tsconfig.json` needs to be updated to ignore `node_modules` and maybe turn on `skipLibChecks`. I'd also take a look at updating outdated dependencies with `npm outdated` and `npm update`

aecorredor commented:
I tried ignoring `node_modules` and that didn't do it. Maybe with `skipLibChecks` it works. I'll try that when I have time. About outdated dependencies, I'll give it a shot too. Thanks.

aecorredor commented:
@mike-engel so even though `skipLibCheck` should be doing the trick, it still fails because `dtslint` does this check:
```js
for (const key in options) { // tslint:disable-line forin
                switch (key) {
                    case "lib":
                    case "noImplicitAny":
                    case "noImplicitThis":
                    case "strictNullChecks":
                    case "strictFunctionTypes":
                    case "esModuleInterop":
                    case "allowSyntheticDefaultImports":
                        // Allow any value
                        break;
                    case "target":
                    case "paths":
                    case "jsx":
                    case "experimentalDecorators":
                    case "noUnusedLocals":
                    case "noUnusedParameters":
                        // OK. "paths" checked further by types-publisher
                        break;
                   default:
                        if (!(key in mustHave)) {
                            throw new Error(`Unexpected compiler option ${key}`);
                        }
                }
            }
```

Since they don't have `skipLibCheck` included the error is thrown.

Also, even if I add 
```js
                    case "skipLibCheck":
                        // OK. Allow skipping type checking of all declaration files (*.d.ts).
                        break;
 ```

The linter still fails and spits out the same errors, so it's kind of doing nothing. I see the error is `Error: Errors in typescript@2.6 for external dependencies:`, so it's using 2.6 which should respect `skipLibCheck`.

mike-engel commented:
@aecorredor Ok, this is now beyond what I can help you with 😄 

We need some sort of maintainer for this repo, but I'm not sure who to ask. @typescript-bot please help. It looks like if there's no activity for a week a maintainer will get pinged but it'd be great not to wait that long.

aecorredor commented:
@ mike-engel ok so I figured this out. `dtslint` allows you to specify which typescript version you want on line 5 of your `index.d.ts`. We had `2.6` and with that one the linting failed for `apollo-utilities`, seems they are using a later version of TS. By updating line 5 to `3.0` everything passes. Also updated `graphql-tools` to latest version, and fixed the linting errors I got for `resolveDependee` and `resolveDependees`, which were the following:
![Screen Shot 2019-07-31 at 2 28 17 PM](https://user-images.githubusercontent.com/9114987/62237924-ea161b00-b39f-11e9-9a34-ee013b2064d9.png)

mike-engel commented:
@aecorredor thanks for working on this! Those changes make sense and look good to me. It's unfortunate you had all the trouble, but I'm glad you were able to work through them 😄

aecorredor commented:
@mike-engel no problem man! glad I was able to help and thanks for making these typings in the first place 🤘🏻
